flink-table/flink-table-common/src/main/java/org/apache/flink/table/api/NoMatchingTableFactoryException.java

```diff
@@ -28,6 +29,7 @@
 import java.util.stream.Collectors;
 
 /** Exception for not finding a {@link TableFactory} for the given properties. */
+@PublicEvolving
 public class NoMatchingTableFactoryException extends RuntimeException {
```

Comment:
I think this one should not be `PublicEvolving`, but perhaps we should rather move it outside the `api` package as it's not referenced by any class in this package, but only by internal classes `TableFactoryService` and `FactoryUtil`. Also, the only reason for this class to be useful from the user point of view is that it can be instance checked on, as the class itself is not adding any method with more informations about the error.

Reply:
> perhaps we should rather move it outside the api package as it's not referenced by any class in this package

The idea is to have APIs live in the `api` package, though, so if it is API (see below), then it's the other stuff that should rather live elsewhere.

> but only by internal classes TableFactoryService and FactoryUtil

`FactoryUtil` is not internal, it is public API. And this is a meaningful exception returned from there, so I think it would be better to declare this part of the API and leave it where it is (since it is more correct and moving classes is breaking).

Reply:
I would rather prefer to move it where `FactoryUtil` in the `factory` package, because this exception is a very specific exception which is used only in a specific context. This package on the other hand has a wider scope and audience. As a counter example, `ValidationException` and `TableException` IMO fit this package since they're broad exceptions used by a wide range of APIs. Same discussion applies for `AmbiguousTableFactoryException` and `CatalogNotExistException`.

Reply:
Upon closer inspection, this exception is actually never propagated to the user from `FactoryUtil`, only from `TableFactoryService`, but that one is internal, so we should be good to move it.